x/wasm-storage/keeper/msg_server.go

```diff
@@ -164,3 +167,55 @@ func (m msgServer) UpdateParams(goCtx context.Context, req *types.MsgUpdateParam
 	}
 	return &types.MsgUpdateParamsResponse{}, nil
 }
+
+// RefundTxFee identifies the last message in the tx based on the exclusivity and
+// uniqueness of commit/reveal txs guaranteed by CommitRevealDecorator ante handler
+// and then issues a refund of the collected fees based on the gas consumed.
+func (m msgServer) RefundTxFee(goCtx context.Context, req *types.MsgRefundTxFee) (*types.MsgRefundTxFeeResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	// Check that the authority is the core contract.
+	if _, err := sdk.AccAddressFromBech32(req.Authority); err != nil {
+		return nil, err
+	}
+	coreContract, err := m.CoreContractRegistry.Get(ctx)
+	if err != nil {
+		return nil, err
+	}
+	if coreContract != req.Authority {
+		return nil, types.ErrInvalidAuthority.Wrapf("expected %s, got %s", coreContract, req.Authority)
+	}
+
+	// Decode the last message in the tx.
+	tx, err := m.txDecoder(ctx.TxBytes())
+	if err != nil {
+		return nil, err
+	}
+	feeTx, ok := tx.(sdk.FeeTx)
+	if !ok {
+		return nil, errorsmod.Wrap(sdkerrors.ErrInvalidRequest, "transaction is not a fee tx")
+	}
+
+	msgs := tx.GetMsgs()
+	msgInfo, ok := utils.ExtractCommitRevealMsgInfo(coreContract, msgs[len(msgs)-1])
+	if !ok {
+		return nil, errorsmod.Wrap(sdkerrors.ErrInvalidRequest, "last message in tx is not a commit or reveal")
+	}
+
+	// Refund the fee if the info submitted by the contract matches the last msg.
+	if msgInfo == fmt.Sprintf("%s,%s,%t", req.DrId, req.PublicKey, req.IsReveal) {
+		// Refund the fee if the gas limit is set within a reasonable range.
+		// Note the fee may be zero in simulation setting.
+		if !feeTx.GetFee().IsZero() && ctx.GasMeter().Limit() < ctx.GasMeter().GasConsumed()*3 {
```

Comment:
If the fee is zero we can skip all other checks right?

Reply:
We now check if the fee is zero in the earlier stage and return without an error if it is zero.